### PR TITLE
web: Build two WASM modules, with/without extensions, load the appropriate one

### PR DESCRIPTION
**Pull request by torokati44**
This includes, "unblocks", and extends https://github.com/ruffle-rs/ruffle/pull/5225.

Also adds dynamic fallback to an additional, non-extension-enabled WebAssembly module, mostly for compatibility with ~poor~ Safari.

~The `simd128` extension is not yet directly utilized by us (it possibly is by some dependencies), but I have a few ideas where to use it, improving video decoding performance significantly.~
**EDIT**: The `simd128` feature should help a lot with the color conversion of both H.263 and VP6 videos since https://github.com/ruffle-rs/h263-rs/pull/15 got merged. The IDCT in the H.263 decoder could also potentially benefit from it with some tweaking, but I haven't found/made any speedups there yet.

This also roughly (but not quite) doubles both the time it takes to build the web version, and the size of all web-based release packages.

~What I don't like is that completely blanking out `RUSTFLAGS` also disables the `web_sys_unstable_apis` config option, which is needed for WebGPU.
We could just brush over this, saying that if browser compatibility is the entire reason for this fallback module, then surely WebGPU is not important. (Even though, Safari does have experimental support for it.)~
**EDIT**: `RUSTFLAGS` is now set locally and properly for both modules, always setting the `web_sys_unstable_apis` config option.

### CODE REVIEW

**Comment by torokati44** [comment]
If someone more experienced in TS finds a way to avoid having to upgrade to `es2022` and enabling the `topLevelAwait` experiment, please let me know!

**Comment by torokati44** [comment]
As this also negatively affects CI times, would be nice if this dual-module machinery could be easily disabled.
Then it could even be off by default perhaps, only enabled for the releases.
Just where to put this switch and how to make it work well?

**Comment by torokati44** [comment]
I'm thinking about adding a custom cargo profile for either (or both) of the wasm builds, to not jeopardize the build cache every time. Although that's a fairly new feature. And for frequent builds during development, using sccache is advisable anyway.
What do you think?

**Comment by Herschel** [comment]
My opinion is that anything in stable is fair game, so I think it's worth giving the custom profiles a try!

**Comment by torokati44** [comment]
It looks like until the linked cargo issue is addressed, we're going to be stuck with setting the different `RUSTFLAGS` in `package.json`... :/
Not really tragic, but not that neat either...

**Comment by torokati44** [comment]
Okay, I think I made up my mind about this for now.
As is now, the first config is all-in (at least with the 4 relevant extensions), and the fallback is pure vanilla.
This way, Chrome and Firefox users get the full benefits of their choice, while the other build is guaranteed to keep working everywhere it did until now. Safari users are not benefited in any way (not even on 15.2, because of SIMD), but they are not worse off either. _(And of course, as a general rule, only Apple can know what's best for them, I surely don't.)_

Later, we could raise the bar on the fallback module as well, starting with extensions that worked on Safari 14.1 already (which is just signExtensions, if it matters), then maybe continuing with those that came in with Safari 15 (so, bulkMemory and saturatedFloatToInt).

If you disagree, I'm all ears. Well, of course, I am, even if you agree...

**Comment by torokati44** [comment]
Does any NPM or shell scripting whiz perhaps have any tips on how we could skip building the vanilla .wasm module on CI, while keeping cross-shell compatibility? :/
It should just quietly not run any of the `...-noext` commands if the `CI` envvar is set.

**Comment by danielhjacobs** [comment]
You could try prepending the ext commands with `[[ -z $CI ]] && `

See https://stackoverflow.com/questions/39619956/short-way-to-run-command-if-variable-is-set, and test it yourself.

`[[ -z $USERNAME ]] && echo "Not defined"` should print "Not defined" on linux, but [[ -z $USERNAM ]] && echo "Not defined"` should print nothing (unless you set USERNAM).

**Comment by torokati44** [comment]
I don't think this works in CMD.EXE as well... :/ That's the hard part. But thanks anyway!

**Comment by danielhjacobs** [comment]
This works on any system with Python:

python -c "import os, sys;sys.exit(1) if 'CI' in os.environ else sys.exit(0)" && echo 'a'

But needing to install python for CI may be asking a bit much...

**Comment by torokati44** [comment]
A Node equivalent of this did come up... But I tried to find a simpler solution.

**Comment by torokati44** [comment]
I think this is ready for review.

File: web/packages/core/src/load-ruffle.ts
Comment (adrian17) [inline web/packages/core/src/load-ruffle.ts:46]: maybe let's not log the happy case, let's not logspam random websites too much :)

File: web/packages/core/src/load-ruffle.ts
Comment (adrian17) [inline web/packages/core/src/load-ruffle.ts:69]: Maybe we could do that in the same spot we do
```
console.log("New Ruffle instance created.");
```
?
This way we logspam a bit less.
For example:
```
New Ruffle instance created (wasm features: on).
New Ruffle instance created (wasm features: off).
```

File: web/packages/core/package.json
Comment (adrian17) [inline web/packages/core/package.json:19]: Not happy about how complex it looks like, but I don't see a nice cross-platform way of writing it either :(

**Review by adrian17 — APPROVED** [review]

**Review by relrelb — CHANGES_REQUESTED** [review]

**Review by Herschel — COMMENTED** [review]
Thank you! This is great, I'm only going to add to the bikeshedding over the naming a bit. `fancy` and `REAL_FANCY` feel slighty too tongue-in-cheek, I'd prefer something more straightforward like `wasm_extensions`.

File: web/packages/core/package.json
Comment (torokati44) [reply web/packages/core/package.json:19]: Honestly, same... :|

File: web/packages/core/package.json
Comment (torokati44) [reply web/packages/core/package.json:27]: Ooh, I wanted to do something similar, just haven't realised that `cross-env-shell` also does platform-independent environment variable substitution! But if this works even in `CMD.EXE` - and according to the docs, it should -, that would be really nice!

File: web/package.json
Comment (torokati44) [reply web/package.json:38]: Now that it is more internal, because users can use the `build:full-release` script instead of setting this directly, perhaps this would be a better name indeed.

File: web/packages/core/package.json
Comment (torokati44) [reply web/packages/core/package.json:16]: Huh, indeed, I have no idea how I haven't thought of this by myself, thanks!

File: web/packages/core/package.json
Comment (torokati44) [reply web/packages/core/package.json:39]: Yep, good point, thanks!

File: web/README.md
Comment (torokati44) [reply web/README.md:86]: Alright, going with "WebAssembly".

**Comment by torokati44** [comment]
> `fancy` and `REAL_FANCY` feel slighty too tongue-in-cheek, I'd prefer something more straightforward like `wasm_extensions`.

That wasn't my intention (_here, at least_...), I just ran out of words that vaguely mean what I wanted to convey, while not being so overloaded already elsewhere in the project as to be really confusing, and are not too long either. Anyway, I already replaced `REAL_FANCY` based on @relrelb's suggestion. And I guess `wasm_extensions` won't eat up too much additional disk space, network bandwidth, or screen area either... :)

File: web/packages/core/package.json
Comment (torokati44) [reply web/packages/core/package.json:19]: Well, except for what @relrelb just commented, using `cross-env-shell`!

File: web/packages/core/src/load-ruffle.ts
Comment (torokati44) [reply web/packages/core/src/load-ruffle.ts:46]: Alright.

File: web/packages/core/package.json
Comment (torokati44) [reply web/packages/core/package.json:21]: Ooh, that's nice to hear, @danielhjacobs! Might still take a while until it gets into stable I believe, but I'll keep an eye out for it!

Comment (torokati44) [reply web/packages/core/package.json:21]: I would have liked something like that as well, @relrelb, but unfortunately, according to https://doc.rust-lang.org/cargo/reference/config.html#buildrustflags:

image

I think the `RUSTFLAGS` envvar overrides anything else, without any concatenation... :/

File: web/packages/core/src/load-ruffle.ts
Comment (torokati44) [reply web/packages/core/src/load-ruffle.ts:69]: Hm, I like the idea, but then I'd either have to return `isWasmSimdUsed` from `fetchRuffle` and then from `loadRuffle`, or make it a member of `Ruffle`. Is that right?

Comment (adrian17) [reply web/packages/core/src/load-ruffle.ts:69]: > or make it a member of Ruffle

Yeah, I don't see anything wrong with that :)

**Comment by torokati44** [comment]
Thanks a lot for the comments and suggestions!
While addressing them, I've had a minor mishap with git, so I can only hope that I haven't left anything out.
If I did, please add notes to the issues still remaining.

File: web/packages/core/package.json
Comment (torokati44) [reply web/packages/core/package.json]: Welp, it doesn't look like this works on Windows... :/

Comment (torokati44) [reply web/packages/core/package.json]: Okay, I got it to work in the end! \o/ Thanks!

**Comment by torokati44** [comment]
Okay, I think I resolved all comments that I could, and brought it again to a presentable state.
So, this is ready for the next round!

File: .github/workflows/release_nightly.yml
Comment (relrelb) [inline .github/workflows/release_nightly.yml:271]: Quite petty, but maybe rename this to `build:full`?

Comment (torokati44) [reply .github/workflows/release_nightly.yml:271]: Hmm, I'm not sure about this one... Feels less descriptive to me, while I'm not seeing how it's any better. What do you think, @adrian17?

Comment (danielhjacobs) [reply .github/workflows/release_nightly.yml:271]: Personally, I lean towards build:release. I understand and somewhat agree with shortening this, but build:release seems to match better with build:debug. As in "this is the debug version. this is the release version". Maybe wait for an opinion from Adrian too though. Calling this the full version, as you mention, feels less descriptive.

Comment (torokati44) [reply .github/workflows/release_nightly.yml:271]: FWIW:
image

So, I am fine with `build:release` as well.

Comment (adrian17) [reply .github/workflows/release_nightly.yml]: > but build:release seems to match better with build:debug
> As in "this is the debug version. this is the release version".

This is my reasoning:
The issue is that these days, "release" means "optimized", not "this is what we use to release the software".

For comparison, these are (very general) build configurations for Firefox:
- max debug, unoptimized; you basically never want to actually run this
- normal optimized build (what `cargo` or Visual Studio would call "release")
- increasingly fancy configurations with LTO, PGO, branding etc - for actual software release or more thorough tests and benchmarks

And these more or less correspond to our configurations:
- max debug, unoptimized; you basically never want to actually run this
- optimized (current `npm run build`)
- optimized with extra debug info/avm tracing (current `npm run build:debug`)
- increasingly fancy configurations with special wasm features

Now, when a dev comes and wants to build it, they most likely want "normal optimized build". But for a Rust programmer, this would probably be `cargo run --release`. But if they apply this logic here and run `build:release`, they get "final-release" and not "development-release" :) Which is why I'd prefer to reserve some more explicit wording for "this is max optimized, not for everyday development".

Comment (danielhjacobs) [reply .github/workflows/release_nightly.yml]: Then pulling from a word you used, what about build:optimized?

Comment (torokati44) [reply .github/workflows/release_nightly.yml]: That isn't perfect either: Even for the plain `build` command now, `-O` is passed to `rustc` and `wasm-opt` is also called on the output either way... :|

Comment (torokati44) [reply .github/workflows/release_nightly.yml]: > Which is why I'd prefer to reserve some more explicit wording for "this is max optimized, not for everyday development".

Although maybe most testing would be done on desktop anyway? Unless it's for a web-specific backend, sure...

Comment (danielhjacobs) [reply .github/workflows/release_nightly.yml]: > That isn't perfect either: Even for the plain build command now, -O is passed to rustc and wasm-opt is also called on the output either way... :|

Yeah, I realized that and deleted my comment suggesting build:optimized. Maybe build:fully-featured, but if the purpose of renaming is to shorten the command, that sort of defeats the purpose.

Comment (torokati44) [reply .github/workflows/release_nightly.yml]: I propose yet another variant: `build:dual-wasm` :| - This one doesn't have `release` it.
And `build:dual-wasm-release` is way too long again?

**Comment by torokati44** [comment]
I get the feeling that the only thing holding this up now is the name of that script. And I am not sure how to resolve this distributed decision, with all options having pros and contras brought up by different contributors. Which is awkward.
Honestly, just having this finally in is more important to me now than this single final detail.
And I don't even know who should have the final say in this bikeshed-esque question.

File: .github/workflows/release_nightly.yml
Comment (torokati44) [reply .github/workflows/release_nightly.yml]: @relrelb Could you please elaborate on what you had against `build:full-release` originally?

Anyway, what do you and @danielhjacobs think about `build:dual-wasm`?

Comment (danielhjacobs) [reply .github/workflows/release_nightly.yml]: I'm fine with dual-wasm

Comment (relrelb) [reply .github/workflows/release_nightly.yml]: I don't have a strong preference to any name (this is a classical example for bikeshedding), I just thought `build:full` is a little more concise and descriptive name. I'm fine with staying on `build:full-release`, changing to `build:dual-wasm`, or even changing to something else.

File: web/package.json
Comment (relrelb) [inline web/package.json:42]: Also very petty, but `--workspaces` is not really needed here (see https://github.com/ruffle-rs/ruffle/pull/6025#discussion_r782749655).

File: web/packages/core/tsconfig.json
Comment (relrelb) [inline web/packages/core/tsconfig.json:5]: Just from curiosity, why `es2020` is needed?

Comment (torokati44) [reply web/packages/core/tsconfig.json:5]: For the `import()` function.

File: web/package.json
Comment (torokati44) [reply web/package.json]: Okay, I have removed it.

**Comment by torokati44** [comment]
So, if nobody has anything against `npm run build:dual-wasm` (vs. `npm run build:full-release` earlier, or any other option that came up), I think this is ready ... again! :D

**Review by relrelb — APPROVED** [review]
LGTM!

**Review by relrelb — APPROVED** [review]
LGTM!

**Review by Herschel — APPROVED** [review]
Let's do it! Thank you!

**Comment by torokati44** [comment]
Woooh yeah, thanks a whole hecking bunch and a half to everyone involved! 🥳

**Pull request merged**
